lib/KBaseReportPy/utils/validation_utils.py

```diff
@@ -0,0 +1,191 @@
+    _require_workspace_id_or_name(params)
+    if not validator.validate(params):
+        raise TypeError(_format_errors(validator.errors, params))
+    return params
+
+
+def _require_workspace_id_or_name(params):
+    """
+    We need either workspace_id or workspace_name, but we don't need both
+    cerberus doesn't have good syntax for that, so we do it manually
+    """
+    if ('workspace_id' not in params) and ('workspace_name' not in params):
+        err = {
+            'workspace_name': ['required without workspace_id'],
+            'workspace_id': ['required without workspace_name']
+        }
+        raise TypeError(_format_errors(err, params))
+    return params
+
+
+def _validate_files(files):
+    """
+    Validate that every entry in `files` contains either a "shock_id" or "path"
+    Raise an exception if any `path` value in `files` is non-existent
+    """
+    def _file_or_dir(path):
+        return os.path.isfile(path) or os.path.isdir(path)
+    for f in files:
+        if ('path' not in f) and ('shock_id' not in f):
+            err = {'path': ['required without shock_id'], 'shock_id': ['required without path']}
+            raise TypeError(_format_errors(err, f))
+        if ('path' in f) and (not _file_or_dir(f['path'])):
+            err = {'path': ['does not exist on filesystem']}
+            raise ValueError(_format_errors(err, f))
+
+
+def _validate_html_index(html_links, index):
+    """
+    Validate that the main file (html_link['name']) is present inside the html directory
+    """
+    if (index is None) or (not html_links):
+        return
+    html_link = html_links[index]
+    if 'path' not in html_link:
+        return
+    # If they passed in a file, we don't need to validate
+    if os.path.isfile(html_link['path']):
+        return
+    # If they passed a directory, check that the 'name' exists as a file inside that dir
+    main_path = os.path.join(html_link['path'], html_link['name'])
+    if not os.path.isfile(main_path):
+        raise ValueError("".join([
+            "For html_links, the 'name' key should be the filename of the ",
+            "main HTML file for the report page (eg. 'index.html'). ",
+            "The 'name' you provided was not found: ",
+            html_link['name']
+        ]))
+
+
+def _format_errors(errors, params):
+    """ Make human-readable error messages from a cerberus validation instance """
+    err_str = ""
+    # Create a bulleted list of each cerberus error message
+    for key in errors:
+        for err_msg in errors[key]:
+            err_str += " * " + key + ": " + err_msg + "\n"
+    return "".join([
+        "KBaseReport parameter validation errors:\n",
+        err_str,
+        "You parameters were:\n",
+        pprint.pformat(params),
+        "\n(View the type spec here: ",
+        # TODO update this URL when it moves and is final:
+        "https://github.com/jayrbolton/KBaseReportPy/blob/master/KBaseReportPy.spec)\n",
```

Comment:
could compile the spec to HTML and link to that instead, like this: http://htmlpreview.github.io/?https://github.com/kbaseapps/ReadsUtils/blob/master/ReadsUtils.html

Reply:
although I'm not sure that linking out to a url in an error message is a great idea. Maybe just say check the docs or something like that.

Reply:
I personally really like to see help URLs in error messages. But I agree it doesn't make sense if the URL is too ephemeral. I'll remove for now.